### ISSUE TITLE
[BERT-BAN] Progress Update

### ISSUE THREAD

**Issue by Nana2929** [Nana2929/vqa-rad-with-bert#3]
## 目標

以 BERT 與 transformer Multi-head Attn 機制取代 glove + LSTM 更新的機制，創造出以 BERT 動態決定 embedding 的效果。

## 調參大法／困難之處
- 沒有完全理解 TypeAttn 的用途。
- 很難調 lr，可能要 BERT 和 BAN 分開調整 lr。

## Quick Overview of What We Have Done

1. Revise `lib/language/language_model.py` 加入 `BERTWordEmbedding `as Question Embedder Model。
2. Revise `lib/language/classify_question.py` 融合 `BERTWordEmbedding` 重新訓練一個 type (open-close) classifier，並存起來當 checkpoint for later use。這部分也會需要改動到 `type_classifier.py` （訓練程式碼）。
4. Revise `lib/dataset/dataset_RAD_bert.py`，將 encode 過程移除（不使用 dictionary），直接在 `BERTWordEmbedding` 中 forward and encode。
5.  Revise `lib/BAN/multi_level_model.py`，移除原本的 `WordEmbedding` 與 `QuestionEmbedding` 並調整 `typeAttn` 和 `q_emb_model` 的呼叫 API。Note: `WordEmbedding` 的功用類似單純 encode 字詞成為 text embedding，`QuestionEmbedding` 接收`WordEmbedding` 的結果後會用 GRU 去更新這些 embeddings 內部的參數。

## Current Architecture (2023.06.03) 
Use BERT final output as embedding. 
During training, BERT will be fine-tuned. 
For TypeAttn, use BERT emb layer and output layer to do Yoshua Bengio attention (`QuestionAttention`) . 
For my own reference, this is using the **refined** image for training. 

### BERTWordEmbedding
```python
# @lib/language/language_model.py
from typing import List
import torch
import torch.nn as nn
from torch.autograd import Variable
import numpy as np
from transformers import AutoModel, AutoTokenizer, AutoConfig


class BERTWordEmbedding(nn.Module):
    """BERT Word Embedding
    """

    def __init__(self, bert_model_name, dropout=0.5, max_length=20):
        super(BERTWordEmbedding, self).__init__()

        self.tokenizer =  AutoTokenizer.from_pretrained(bert_model_name)
        self.config = AutoConfig.from_pretrained(bert_model_name)
        self.model = AutoModel.from_pretrained(bert_model_name, config=self.config)

        self.max_length = max_length
        self.output_layer = nn.Linear(self.config.hidden_size, 1024)
        self.vocab_size = self.config.vocab_size
        self.emb_dim = self.config.hidden_size
        # no longer used but the pretrained checkpoint requires this
        self.emb = nn.Embedding(self.vocab_size + 1,
                                self.emb_dim,
                                padding_idx=self.find_padding_idx())
        self.emb.weight.requires_grad = True # fixed
       
    def find_padding_idx(self):
        pad_id = self.tokenizer.convert_tokens_to_ids(['[PAD]'])[0]
        return pad_id

    # Archive
    def _prepare_last_4(self, x: str):
        """
        x: a string/question
        """

        tokenized_text = self.tokenizer.tokenize(x)
        indexed_tokens = self.tokenizer.convert_tokens_to_ids(tokenized_text)
        segments_ids = [1] * len(indexed_tokens)
        tokens_tensor = torch.tensor([indexed_tokens])
        segments_tensors = torch.tensor([segments_ids])
        with torch.no_grad():
            encoded_layers, _ = self.model(tokens_tensor, segments_tensors)
        token_embeddings = torch.stack(encoded_layers, dim=0)
        token_embeddings = torch.squeeze(token_embeddings, dim=1)


        # Stores the token vectors, with shape [22 x 3,072]
        token_vecs_sum = []
        # `token_embeddings` is a [12 x 768] tensor.

        # For each token in the sentence...
        for token in token_embeddings:
            sum_vec = torch.sum(token[-4:], dim=0)
            # Use `cat_vec` to represent `token`.
            token_vecs_sum.append(sum_vec)
        return token_vecs_sum  # shape [seq_len, 768]

    def is_bert_training(self):
        return self.model.training

    def fully_forward(self, input_ids, token_type_ids=None, attention_mask=None, **kwargs):
        # (batch_size, seq_len, 1024)
        return self.output_layer(self.model(input_ids, token_type_ids, attention_mask))

    def encode_and_cast_dim_forward(self, questions:List[str]):
        # (batch_size, seq_len, 1024)
        # 1024: HID_DIM
        return  self.output_layer(self.encode_and_forward(questions)) # (batch_size, seq_len, 1024)  as type attention

    def emb_forward(self, questions: List[str]):
        # use first layer output as embedding
        device = next(self.model.parameters()).device
        tokenized = self.tokenizer.batch_encode_plus(questions,
                                                  return_tensors="pt",
                                                  max_length=self.max_length,
                                                  truncation=True,
                                                  padding='max_length')
        tokenized = {k: v.to(device) for k, v in tokenized.items()}
        # use BERT embedding layer (0th layer) output
        embeddings = self.model.embeddings(tokenized['input_ids'],
                                           tokenized['attention_mask'])
        # (batch_size, seq_len, 768)

        return embeddings

    def encode_and_forward(self, questions:List[str]):
        # use last layer output as embedding
        # (batch_size, seq_len, 768)

        device = next(self.model.parameters()).device
        encode = self.tokenizer.batch_encode_plus(questions,
                                                  return_tensors="pt",
                                                  max_length=self.max_length,
                                                  truncation=True,
                                                  padding='max_length'
                                                  )

        encode = {k: v.to(device) for k, v in encode.items()}
        # https://huggingface.co/docs/transformers/v4.29.1/en/main_classes/output#transformers.modeling_outputs.BaseModelOutputWithPoolingAndCrossAttentions
        logit = self.model(**encode).last_hidden_state # (batch_size, seq_len, 768)
        return logit
```

### Type Attention
```python
# @ lib/language/classify_question.py
class typeAttention(nn.Module):
    def __init__(self, bert_model_name: str,
                 in_dim= 768):
        super(typeAttention, self).__init__()
        self.q_emb_model = BERTWordEmbedding(
            bert_model_name = bert_model_name,
        )

        self.q_attn= QuestionAttention(
            in_dim = in_dim,
            hid_dim = 1024)
        self.f_proj = linear(in_dim, 1024)
        self.f_fc1 = linear(1024, 2048)
        self.f_fc2 = linear(2048, 1024)
        self.f_fc3 = linear(1024, 1024)

    def forward(self, question):
        q_first = self.q_emb_model.emb_forward(question)
        q_last = self.q_emb_model.encode_and_forward(question)
        # q_first shape: torch.Size([16, 20, 768]) = [b, max_seq_len, emb_dim] (emb layer of BERT)
        # q_last shape: torch.Size([16, 20, 768]) = [b, max_seq_len, emb_dim]   (output layer of BERT)
        q_final = self.q_attn(q_first, q_last)
        # q_final shape: torch.Size([16, 768]) = [b, emb_dim]
        q_proj = self.f_proj(q_final)
        # q_proj shape: torch.Size([16, 1024]) = [b, hid_dim]

        x_f = self.f_fc1(q_proj)
        x_f = F.relu(x_f)
        x_f = self.f_fc2(x_f)
        x_f = F.dropout(x_f)
        x_f = F.relu(x_f)
        x_f = self.f_fc3(x_f)

        return x_f
```

## Config
```yaml
NAME: 'QCR.CLIPRN50.AE.ROCO.VQARAD.16batchsize.200epoch.withTFIDF.nondeterministic'
OUTPUT_DIR: './output/qcr/pubmedclipRN50_ae/roco/VQARAD'
CPU_MODE: False
SEED: 88

DATASET:
  DATASET: "RAD"
  DATA_DIR: "./data_rad"
  DATA_TYPE: 'jpg'
  EMBEDDER_MODEL: 'bert-base-uncased'
  EMB_DIM: 768

LOSS:
  LOSS_TYPE: 'BCELogits'

TRAIN:
  BATCH_SIZE: 12
  N_EPOCH: 200
  NUM_WORKERS: 2
  RESUME: False
  INPUT_SNAPSHOT: ""
  OPTIMIZER:
    TYPE: 'ADAMX'
    BASE_LR: 1e-3
    MOMENTUM_CNN: 0.05
    EPS_CNN: 1e-5
  ACTIVATION: 'relu'
  DROPOUT: 0.5
  ATTENTION:
    MODE: 'BAN'
    GLIMPSE: 10  # Attn heads
    USE_COUNTER: False
    NUM_STACKS: 2    # For SAN
  QUESTION:
    RNN: 'GRU' # not used
    LENGTH: 20
    TFIDF: False
    CAT: False # not used
    HID_DIM: 1024   ## Dim of joint semantic features
  VISION:
    V_DIM: 1088    ## Visual input dim : 1024 + 64
    AUTOENCODER: True
    AE_PATH: "pretrained_ae.pth"
    AE_ALPHA: 0.001
    MAML: False
    CLIP: True
    CLIP_PATH: "/home/nanaeilish/projects/mis/Medical-VQA/PubMedCLIP_RN50.pth"
    CLIP_VISION_ENCODER: "RN50"
    OTHER_MODEL: False


TRANSFORMS:
  TRAIN_TRANSFORMS: ("random_resized_crop", "random_horizontal_flip")
  TEST_TRANSFORMS: ("shorter_resize_for_crop", "center_crop")

TEST:
  BATCH_SIZE: 8
  NUM_WORKERS: 4
  MODEL_FILE: "./output/qcr/pubmedclipRN50_ae/roco/VQARAD/QCR.CLIPRN50.AE.ROCO.VQARAD.16batchsize.200epoch.withTFIDF.nondeterministic/62_best.pth"
  RESULT_DIR: "./results"
```
## Verify if BERT is being fine-tuned with VQA loss  
A: Yes 

```python
q_emb = self.q_emb_model.encode_and_cast_dim_forward(q) # [batch, q_len, q_dim], where q_dim is TRAIN.QUESTION.HID_DIM
        print('Is q_emb_model training?', self.q_emb_model.is_bert_training())
```

![Image](https://user-images.githubusercontent.com/58811089/243030703-ed9f5667-7312-42b0-a0f3-2ae03fd3a0bc.png)
